### PR TITLE
Feature/cppstd tools

### LINKED ISSUE #6687: [question] cppstd helper utilities
As per contribution guide I am first asking question, only then drafting a pull request. 

I was thinking of slightly reworking functionality in `tools/settings.py` - this file has functionalities around cppstd. 

What I would see beneficial:

1) Add mirroring function for 'max' cppstd.
2) Rework code so that deduced_cppstd is public
3) Add method like.. `deduced_cppstd_long`  which returns cppstd in form of '1998', '2011', so that comparision in recipes is easier for exampel:
if `cppstd > 11` is now tricky because you have to keep mind of '98' and also you have to remember that cppstd defaults to None if not specified in the profile.

What do you guys think?
Can I go ahead and create a pr for this?

- [X] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).

### PR DESCRIPTION
**Pull request by Minimonium**
Changelog: (Feature): Add tools for `cppstd` manipulation and general checks
Docs: https://github.com/conan-io/docs/pull/XXXX

Closes #6687

If required, I can break this PR into multiple smaller PRs.

- [x] Refer to the issue that supports this Pull Request.
- [x] If the issue has missing info, explain the purpose/use case/pain/need that covers this Pull Request.
- [x] I've read the [Contributing guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).
- [x] I've followed the PEP8 style guides for Python code.
- [ ] I've opened another PR in the Conan docs repo to the ``develop`` branch, documenting this one. 

<sup>**Note:** By default this PR will skip the slower tests and will use a limited set of python versions. Check [here](https://github.com/conan-io/conan/blob/develop/.github/PR_INCREASE_TESTING.md) how to increase the testing level by writing some tags in the current PR body text.</sup>

### CODE REVIEW

**Comment by Minimonium** [comment]
List of additions:
* `deduced_cppstd` - returns either the specified or default compiler version. If there is not information about default compiler version - returns `None`.
* `normalized_cppstd` - returns standard version in the form of a year. Removes extensions.
* `check_gnu_extension` - Raises an exception if a given standard version doesn't contain the `gnu` prefix.
* `check_max_cppstd`/`valid_max_cppstd` are the upper bound to the min versions. I'm not sure if we require them if we also add more generalized things.
* `check_cppstd` and `check_compiler` are tools that intended to solve the https://github.com/conan-io/conan-center-index/issues/54

**Comment by Minimonium** [comment]
An additional function that _requires_ this PR - `compatible_cppstd`. It fills the `compatible_packages` field with `cppstd` compatible structures according to specified behaviour. By default, it's forward, but not backward, compatible.

File: conans/client/tools/settings.py [hunk 1]
```diff
@@ -1,77 +1,382 @@
...
+    return str(_cppstd).replace("gnu", "")
+
+
+def _contains_extension(cppstd):
+    """ Returns if cppstd contains gnu extension
+
+    :param cppstd: cppstd version
+    :return: True, if current cppstd contains the gnu extension. Otherwise, False.
+    """
+    return str(cppstd).startswith("gnu") and _remove_extension(cppstd).isdigit()
```
Comment (SSE4) [inline conans/client/tools/settings.py:49]: why is digit? what about 0x, 2a, 1z and so on?

File: conans/client/tools/settings.py [hunk 2]
```diff
@@ -1,77 +1,382 @@
...
+        a ConanUnknownConfiguration exception will be raised.
+
+    :param conanfile: ConanFile instance
+    :param min: Minimal cppstd version required
+    :param max: Maximal cppstd version required
+    :param excludes: A list of cppstd version excluded
+    :param gnu_extensions: GNU extension is required (e.g gnu17)
+    :param strict: Unkown configurations are invalid
+    """
+    if min and not str(min).isdigit():
```
Comment (SSE4) [inline conans/client/tools/settings.py:189]: I don't think we need so many so strict type checks everywhere (given the duck typing, it's not a very pythonic way for my point of view).

File: conans/client/tools/settings.py (hunk 1 above)
Comment (Minimonium) [reply conans/client/tools/settings.py:49]: `cppstd` never contain 0x, 2a, 1z. `cppstd` - the value in settings

File: conans/client/tools/settings.py (hunk 2 above)
Comment (Minimonium) [reply conans/client/tools/settings.py:189]: The point is that these are user-facing functions and users are not necessary familiar with the Python philosophy. I'm not necessary against removing them, but I'd prefer to have clearer error messages for users other than a python stack trace.

File: conans/client/tools/settings.py (hunk 1 above)
Comment (SSE4) [reply conans/client/tools/settings.py:49]: nice!

File: conans/client/tools/settings.py [hunk 3]
```diff
@@ -1,77 +1,382 @@
...
-            raise ConanInvalidConfiguration("The cppstd GNU extension is required")
+        In case the current cppstd doesn't fit the maximal version required
+        by cppstd, a ConanInvalidConfiguration exception will be raised.
 
-    def deduced_cppstd():
-        cppstd = conanfile.settings.get_safe("compiler.cppstd")
-        if cppstd:
-            return cppstd
+    :param conanfile: ConanFile instance with cppstd information
+    :param cppstd: Minimal cppstd version required
```
Comment (uilianries) [inline conans/client/tools/settings.py]:
```suggestion
    :param cppstd: Maximum cppstd version required
```

File: conans/client/tools/settings.py (hunk 2 above)
Comment (uilianries) [reply conans/client/tools/settings.py:189]: usually I don't like many checks, but as those methods a public for users, they can pass anything, so I agree adding these messages.

File: conans/client/tools/settings.py [hunk 4]
```diff
@@ -1,77 +1,382 @@
...
+    """ Check if the current compiler fits specified requirements
+
+        In case the current compiler doesn't fit specified requirements
+        a ConanInvalidConfiguration exception will be raised.
+
+        In case information about compiler is lacking and strict flag is set
+        a ConanUnknownConfiguration exception will be raised.
+
+    :param conanfile: ConanFile instance
+    :param required: A dict of required compiler versions required
```
Comment (uilianries) [inline conans/client/tools/settings.py]:
```suggestion
    :param required: A dict of required compiler versions required e.g. {"gcc": "9"}
```

File: conans/client/tools/settings.py [hunk 5]
```diff
@@ -1,77 +1,382 @@
...
+        if not current_cppstd:
+            raise ConanInvalidConfiguration(
+                "Could not detect default cppstd for the current compiler.")
+    else:
+        current_cppstd = str(current_cppstd)
+
+    # We have a list of original cppstd setting values because we'll
+    # need it to pass to compatible packages
+    # FIXME: Conanfile should carry range information of a setting with it
+    base_cppstd_range = ["98", "11", "14", "17", "20"]
```
Comment (uilianries) [inline conans/client/tools/settings.py:310]: can we share it from settings? For now we have init file, which contains the yaml content, but we could move the cppstd list and re-use it for setting.yml and here.

Comment (Minimonium) [reply conans/client/tools/settings.py:310]: That's exactly what my comment in the `FIXME` suggests, but it sounds like it'd be better to do it in a separate PR.

Comment (Minimonium) [reply conans/client/tools/settings.py:310]: To elaborate, one of the numerous structures that can carry settings actually contains the range field which contains all the values we'd need for a given setting. The problem is that it's used inconsistently across the codebase, so it requires a cleanup.

**Comment by memsharded** [comment]
It seems this PR went outdated, and now it is targeting the old 1.X, and conflicting. New tools should be in the ``from tools.xxx`` scope now. I am closing this, I think it would be better to open a new PR if necessary. Thanks very much again for your contribution!

**Pull request closed without merging**